plutus-core/plutus-core/src/PlutusCore/Value.hs

```diff
@@ -203,39 +287,40 @@ instance Pretty Value where
 the size of the largest inner map.
 -}
 insertCoin :: ByteString -> ByteString -> Integer -> Value -> BuiltinResult Value
-insertCoin currency token amt v@(Value outer sizes size neg)
-  | amt == 0 = pure $ deleteCoin currency token v
-  | otherwise = case (k currency, k token) of
-      (Nothing, _) -> fail $ "insertCoin: invalid currency: " <> show (B.unpack currency)
-      (_, Nothing) -> fail $ "insertCoin: invalid token: " <> show (B.unpack token)
-      (Just ck, Just tk) ->
+insertCoin unsafeCurrency unsafeToken unsafeAmount v@(Value outer sizes size neg)
+  | unsafeAmount == 0 = pure $ deleteCoin unsafeCurrency unsafeToken v
+  | otherwise = case (k unsafeCurrency, k unsafeToken, quantity unsafeAmount) of
+      (Nothing, _, _) -> fail $ "insertCoin: invalid currency: " <> show (B.unpack unsafeCurrency)
+      (_, Nothing, _) -> fail $ "insertCoin: invalid token: " <> show (B.unpack unsafeToken)
+      (_, _, Nothing) -> fail $ "insertCoin: quantity out of bounds: " <> show unsafeAmount
+      (Just currency, Just token, Just qty) ->
```

Comment:
Have you heard of this thing called `Monad`? :slightly_smiling_face: 

You can just make `k` and `quantity` run in `BuiltinResult` and perform all these checks using tidy do-notation.

Maybe we should distinguish between `currency` and `token` statically too,

Anyway, it's fine as is, just nitpicking.

Although I'm not a fan of the pointlessly lazy `Maybe`.